3p/ampcontext.js

```diff
@@ -0,0 +1,182 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import './polyfills';
+import {dev, user} from '../src/log';
+import {IframeMessagingClient} from './iframe-messaging-client';
+
+/**
+   Enum for the different postmessage types for the window.context
+   postmess api.
+*/
+export const MessageType_ = {
+  SEND_EMBED_STATE: 'send-embed-state',
+  EMBED_STATE: 'embed-state',
+  SEND_EMBED_CONTEXT: 'send-embed-context',
+  EMBED_CONTEXT: 'embed-context',
+  SEND_INTERSECTIONS: 'send-intersections',
+  INTERSECTION: 'intersection',
+  EMBED_SIZE: 'embed-size',
+  EMBED_SIZE_CHANGED: 'embed-size-changed',
+  EMBED_SIZE_DENIED: 'embed-size-denied',
+};
+
+export class AmpContext extends IframeMessagingClient {
+
+  /**
+   *  @param {Window} win The window that the instance is built inside.
+   */
+  constructor(win) {
+    super(win);
+    this.setupMetadata_();
+
+    /** Calculate the hostWindow / ampWindow_ */
+    const sentinelMatch = this.sentinel.match(/((\d+)-\d+)/);
+    dev().assert(sentinelMatch, 'Incorrect sentinel format');
+    this.depth = Number(sentinelMatch[2]);
+    const ancestors = [];
+    for (let win = this.win_; win && win != win.parent; win = win.parent) {
+      // Add window keeping the top-most one at the front.
+      ancestors.push(win.parent);
+    }
+    ancestors.reverse();
+
+    /** @private */
+    this.ampWindow_ = ancestors[this.depth];
+  }
+
+  /** @override */
+  getHostWindow() {
+    return this.ampWindow_;
+  }
+
+  /** @override */
+  getSentinel() {
+    return this.sentinel;
+  }
+
+  /** @override */
+  registerCallback_(messageType, callback) {
+    user().assertEnumValue(MessageType_, messageType);
```

Comment:
This allows the `send-*` types, which weren't allowed before. Is that intentional?